File: leetcode/0516.Longest-Palindromic-Subsequence/LongestPalindromicSubsequence.go
package main

import "fmt"

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Runtime: 32 ms, faster than 66.18% of Go online submissions for Longest Palindromic Subsequence.
func longestPalindromeSubseq(s string) int {
	strLen := len(s)
	dp := make([][]int, strLen)
	for i := 0; i < strLen; i++ {
		dp[i] = make([]int, strLen)
		dp[i][i] = 1
	}

	// i -> 0 and j -> strLen
	for i := strLen; i >= 0; i-- {
		for j := i + 1; j < strLen; j++ {
			if s[i] == s[j] {
				dp[i][j] = dp[i+1][j-1] + 2
			} else {
				dp[i][j] = max(dp[i][j-1], dp[i+1][j])
			}
		}
	}
	return dp[0][strLen-1]
}

func isReversedString(s string, strLen int) bool {
	for i, j := 0, strLen-1; i < j; {
		if s[i] != s[j] {
			return false
		}
		i++
		j--
	}
	return true
}

// Early Return
// Runtime: 8 ms, faster than 97.06% of Go online submissions for Longest Palindromic Subsequence.
func longestPalindromeSubseqV2(s string) int {
	strLen := len(s)
	if isReversedString(s, strLen) {
		return strLen
	}
	dp := make([][]int, strLen)
	for i := 0; i < strLen; i++ {
		dp[i] = make([]int, strLen)
		dp[i][i] = 1
	}

	// i -> 0 and j -> strLen
	for i := strLen; i >= 0; i-- {
		for j := i + 1; j < strLen; j++ {
			if s[i] == s[j] {
				dp[i][j] = dp[i+1][j-1] + 2
			} else {
				dp[i][j] = max(dp[i][j-1], dp[i+1][j])
			}
		}
	}
	return dp[0][strLen-1]
}

func main() {
	fmt.Println(longestPalindromeSubseq("aabba"))
}

File: leetcode/0516.Longest-Palindromic-Subsequence/LongestPalindromicSubsequence.py
class Solution:

    # Runtime: 496 ms, faster than 98.41% of Python3 online submissions for Longest Palindromic Subsequence.
    def longestPalindromeSubseq(self, s: str) -> int:
        n = len(s)
        if s == s[::-1]:
            # without this early return section, Runtime will extend to 1228 ms
            # Runtime: 1228 ms, faster than 81.11% of Python3 online submissions for Longest Palindromic Subsequence.
            return n

        dp = [[0]*n for _ in range(n)]
        for i in range(n):
            dp[i][i] = 1

        # final: i --> 0 and  j --> n
        # xxxxxxxxxx
        #      i  j
        for i in range(n-2, -1, -1):
            for j in range(i+1, n):
                if s[i] == s[j]:
                    dp[i][j] = dp[i+1][j-1] + 2
                else:
                    dp[i][j] = dp[i+1][j] if dp[i+1][j] > dp[i][j-1] else dp[i][j-1]
                        
        import pprint
        print(s)
        pprint.pprint(dp)
        return dp[0][-1]

if __name__ == "__main__":
    print(Solution().longestPalindromeSubseq("abxba"))
